zookeeper-server/src/test/java/org/apache/zookeeper/server/quorum/QuorumPeerReadOnlyTest.java

```diff
@@ -0,0 +1,172 @@
+            mt[i] = new MainThread(i, clientPorts[i], quorumCfgSection + "\nclientPort=" + clientPorts[i]);
+            mt[i].start();
+            
+        }
+        
+        for (int i = 0; i < SERVER_COUNT; i++) {
+            zk[i] = new ZooKeeper("127.0.0.1:" + clientPorts[i], ClientBase.CONNECTION_TIMEOUT, this, true);
+        }
+        
+        waitForAll(zk, States.CONNECTED);
+        
+        zk[0].create("/test", "Test".getBytes(), Ids.OPEN_ACL_UNSAFE, CreateMode.PERSISTENT);
+
+        // Shutdown all clients first, to ensure that the close zxid is seen and logged by all servers
+        for (int i = 0; i < SERVER_COUNT; i++) {
+            zk[i].close();
+        }
+        // Shutdown all servers, before starting one again, as the commonly used data storage is cleaned
+        // up, when the first server is stopped
+        for (int i = 0; i < SERVER_COUNT; i++) {
+            mt[i].shutdown();
+        }
+        
+        mt[SERVER_COUNT-1].start();
+        // Ensure that close will trigger a timeout
+        zk[SERVER_COUNT-1] = new ZooKeeper("127.0.0.1:" + clientPorts[SERVER_COUNT-1], ClientBase.CONNECTION_TIMEOUT, this, true) {
+//            @Override
+            public void close2() {
+                getTestable().injectSessionExpiration();
+                cnxn.disconnect();
+                try {
+                    Thread.sleep(ClientBase.CONNECTION_TIMEOUT * 2);
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                }
+            }
+        };
+        
+        waitForAll(new ZooKeeper[] {zk[SERVER_COUNT-1]}, States.CONNECTEDREADONLY);
+        
+        // Read some data
+        assertTrue(Arrays.equals(zk[SERVER_COUNT-1].getData("/test", false, null), "Test".getBytes()));
+        
+        // Kill the client, without notification
+        zk[SERVER_COUNT-1].close();
+       
+        // Start enough server for a quorum, but not all yet. Don't start any client, to avoid changing the zxid
+        for (int i=SERVER_COUNT-1; i >= ((SERVER_COUNT-1)/2); i--) {
+            if (i != SERVER_COUNT-1) {
+                mt[i].start();
+            }
+        }
+        ensureLeaderIsPresent(mt);
+
+        // Seems like without a waiting time, the test does succeeded sometimes without the fix applied
+        Thread.sleep(1000);
+        
+        // Start the remaining servers
+        for (int i = (SERVER_COUNT-1)/2-1; i >= 0; i--) {
+            mt[i].start();
+        }
+        ensureRemainingserversPresent(mt);
+
+        for (int i = 0; i < SERVER_COUNT; i++) {
+            zk[i] = new ZooKeeper("127.0.0.1:" + clientPorts[i], ClientBase.CONNECTION_TIMEOUT, this, true);
+        }
+        waitForAll(zk, States.CONNECTED);
+
+        // Ensure that all clients can read the data
+        for (int i = 0; i < SERVER_COUNT; i++) {
+            assertTrue(Arrays.equals(zk[i].getData("/test", false, null), "Test".getBytes()));
+        }
+        
```

Comment:
For better consistency check, I suggest that we can create multiply znodes(due to `TRUNC`), then assert the znode count and data content.

Reply:
I am not 100% sure what you mean with this point.
I've appended now a couple of nodes, which are created at the beginning and checked at the end.
Adding data in between does hide the problem, iirc. At least if it is done after the quorum was able to restore, before all server do show up.

Reply:
IIRC, I reproduced this issue by finding the loss of some children nodes. Essentially, they're all update loss by `TRUNC`